packages/api/src/Base.ts

```diff
@@ -338,122 +339,115 @@ export default abstract class ApiBase<OnCall> implements ApiBaseInterface<OnCall
 
       result[sectionName] = Object.keys(section).reduce((result, methodName) => {
         result[methodName] = this.decorateExtrinsicEntry(section[methodName], onCall);
 
         return result;
-      }, {} as SubmittableModuleExtrinsics<T>);
+      }, {} as SubmittableModuleExtrinsics<C, S>);
 
       return result;
-    }, {} as SubmittableExtrinsics<T>);
+    }, {} as SubmittableExtrinsics<C, S>);
   }
 
-  private decorateExtrinsicEntry<T> (
-    method: MethodFunction,
-    onCall: (method: OnCallFunction<Observable<Codec | undefined | null>>, params: Array<any>, isSubscription?: boolean) => T
-  ): SubmittableExtrinsicFunction<T> {
-    const decorated: any = (...args: Array<any>): SubmittableExtrinsic<T> =>
-      new SubmittableExtrinsic(this._rx as ApiInterface$Rx, onCall, method(...args));
+  private decorateExtrinsicEntry<C, S> (method: MethodFunction, onCall: OnCallDefinition<C, S>): SubmittableExtrinsicFunction<C, S> {
+    const decorated: any = (...params: Array<CodecArg>): SubmittableExtrinsic<C, S> =>
+      new SubmittableExtrinsic(this._rx as ApiInterface$Rx, onCall, method(...params));
 
-    return this.decorateFunctionMeta(method, decorated) as SubmittableExtrinsicFunction<T>;
+    return this.decorateFunctionMeta(method, decorated) as SubmittableExtrinsicFunction<C, S>;
   }
 
-  private decorateStorage<T> (
-    storage: Storage,
-    onCall: (method: OnCallFunction<Observable<Codec | undefined | null>>, params: Array<any>, isSubscription?: boolean) => T
-  ): QueryableStorage<T> {
-
+  private decorateStorage<C, S> (storage: Storage, onCall: OnCallDefinition<C, S>): QueryableStorage<C, S> {
     return Object.keys(storage).reduce((result, sectionName) => {
       const section = storage[sectionName];
 
       result[sectionName] = Object.keys(section).reduce((result, methodName) => {
         result[methodName] = this.decorateStorageEntry(section[methodName], onCall);
 
         return result;
-      }, {} as QueryableModuleStorage<T>);
+      }, {} as QueryableModuleStorage<C, S>);
 
       return result;
-    }, {} as QueryableStorage<T>);
+    }, {} as QueryableStorage<C, S>);
   }
 
-  private decorateStorageEntry<T> (
-    method: StorageFunction,
-    onCall: (method: OnCallFunction<Observable<Codec | undefined | null>>, params: Array<any>, isSubscription?: boolean) => T
-  ): QueryableStorageFunction<T> {
-    const decorated: any = (...args: any): T =>
-      onCall(
-        (...args: any) => this._rpcRx.state
-          .subscribeStorage([[method, args[0]]])
+  private decorateStorageEntry<C, S> (method: StorageFunction, onCall: OnCallDefinition<C, S>): QueryableStorageFunction<C, S> {
+    // These signatures are allowed and exposed here -
+    //   (arg?: CodecArg): CodecResult;
+    //   (arg: CodecArg, callback: CodecCallback): SubscriptionResult;
+    //   (callback: CodecCallback): SubscriptionResult;
+    const decorated = ((...args: Array<any>): C | S => {
```

Comment:
imo `decorated` should have a better type, its type is the type the end-user would see. So exactly the type you put in the comments, but inside an interface